UM-USElab/gradecraft-development
# 1198 autosave student submissions

### Status
Ready

### Description

![screen shot 2016-12-07 at 2 10 43 pm](https://cloud.githubusercontent.com/assets/17604722/20982531/f96d072a-bc86-11e6-92ba-ebdf091334ce.png)

This PR describes a series of changes that were made to the `Submission` model in order to allow for the saving of draft content, specifically `text_comment`. The newly added field, `text_comment_draft`, is intended for this purpose.

For this PR, a submission draft will be referred to as a `Submission` with `text_comment = nil`, `link = nil`, `submitted_at = nil`, with no `submission_files`. This method has been defined on the submission model as `draft?`.

To make the new submission draft concept more apparent, there is a SubmissionProctor that was added for handling `viewable?` status.

The endpoints as described in `API::Assignments::SubmissionsController` are meant for the checking of saved content on page load as well as the creating and updating of draft submissions.

Because the affected submission form was not originally written in Angular, there are various Angular components that were added that are managed by an instance of the `StudentSubmissionService`. Text content is saved on a change event after a 3.5sec debounce and can also be manually triggered by clicking on the Save Draft button.

### Related PRs
An migration included in this PR closes #1698 

### Deploy Notes
Two migrations exist: One for adding the `text_comment_draft` field to the `Submission` table and one for adding a unique constraint, limiting submissions to one and only one per `Student` and `Assignment`.

In order for the latter to run, preexisting submission data must be cleaned up such that these conditions are met.

### Migrations
Yes

### Steps to Test or Reproduce
Assume that this PR affects all places where Submissions are referenced.

- Ensure that a submission can be submitted with and without a preexisting draft
- Verify that any submission stats are not taking submission drafts into account
- Check that submission exports do not return submission drafts
- Verify that instructors are not be able to view draft content
- Verify that all group members on a group submission are be able to view/edit autosaved content
- Ensure mail notifications are sent under the proper conditions

### Impacted Areas in Application
Again, assume that this PR affects all places where Submissions are referenced. One should expect to be able to properly create, edit, and delete submissions from the app both as an instructor and as a student where required.
